### DISCUSSION TITLE
2020 Seasonal Players Stats

### DISCUSSION THREAD

**Discussion by jcordell** [cwendt94/espn-api#131] [ESPN API] [answered]
Hi, I just started using your package and it looks like it has already implements tons of what I'm trying to do (import rosters, free agents, etc) so it's already super helpful.

I'm trying to get a player's current aggregated season stats.  For example, Christian McCaffery's looks like:

```
{0: {'projected_points': 334.94, 'projected_breakdown': {'23': 236.0222885, 'rushingYards': 1060.346907, 'rushingTouchdowns': 8.319737486, 'rushing2PtConversions': 0.291355662, '27': 212.0, '28': 106.0, '29': 53.0, '30': 42.0, '31': 21.0, '33': 47.0, '34': 23.0, '35': 0.492184603, '36': 0.344529222, '37': 3.093827267, '38': 0.106, '39': 4.492571075, '40': 75.06880755, 'receivingYards': 741.4144806, 'receivingTouchdowns': 2.649136792, 'receiving2PtConversions': 0.14288397, '45': 0.084233715, '46': 0.055046733, '47': 148.0, '48': 74.0, '49': 37.0, '50': 29.0, '51': 14.0, 'receivingReceptions': 90.17001097, '54': 18.0, '55': 9.0, '56': 1.262449485, '57': 0.039, '58': 110.1514268, '60': 8.222406458, '61': 52.48952075, '62': 0.434239632, '63': 0.00966, '66': 1.588375623, '67': 0.606823399, '68': 2.195199021, '70': 0.730652786, '71': 0.339821103, 'lostFumbles': 1.07047389, '73': 1.07047389, '210': 14.125}, 'points': 471.2, 'breakdown': {'0': 2.0, '2': 2.0, '23': 287.0, 'rushingYards': 1387.0, 'rushingTouchdowns': 15.0, 'rushing2PtConversions': 1.0, '27': 270.0, '28': 131.0, '29': 61.0, '30': 49.0, '31': 20.0, '32': 6.0, '33': 48.0, '34': 21.0, '35': 4.0, '36': 3.0, '37': 6.0, '39': 4.83275261, '40': 86.6875, '41': 116.0, 'receivingYards': 1005.0, 'receivingTouchdowns': 4.0, '47': 195.0, '48': 93.0, '49': 43.0, '50': 32.0, '51': 12.0, '52': 2.0, 'receivingReceptions': 116.0, '54': 17.0, '55': 5.0, '56': 2.0, '58': 141.0, '59': 1007.0, '60': 8.6637931, '61': 62.8125, '62': 1.0, '66': 1.0, '68': 1.0, '108': 2.0, '109': 2.0, 'TeamWin': 5.0, '156': 11.0, '158': 116.0, '179': 10.0, '180': 1.0, '183': 2.0, '184': 1.0, '185': 1.0, '210': 16.0}}, 1: {'points': 28.5, 'breakdown': {'23': 23.0, 'rushingYards': 97.0, 'rushingTouchdowns': 2.0, '27': 19.0, '28': 9.0, '29': 4.0, '30': 3.0, '31': 1.0, '33': 4.0, '34': 2.0, '39': 4.217, '40': 97.0, '41': 3.0, 'receivingYards': 38.0, '47': 7.0, '48': 3.0, '49': 1.0, '50': 1.0, 'receivingReceptions': 3.0, '58': 4.0, '59': 35.0, '60': 12.667, '61': 38.0, '156': 1.0, '158': 12.0, '179': 2.0, '210': 1.0}, 'projected_points': 24.62, 'projected_breakdown': {'23': 17.36454637, 'rushingYards': 78.84316424, 'rushingTouchdowns': 0.63923498, 'rushing2PtConversions': 0.0218, '27': 15.0, '28': 7.0, '29': 3.0, '30': 3.0, '31': 1.0, '33': 3.0, '34': 1.0, '35': 0.037, '36': 0.0259, '37': 0.272, '38': 0.00929, '39': 4.540467834, '40': 78.84316424, 'receivingYards': 54.24683453, 'receivingTouchdowns': 0.201719963, 'receiving2PtConversions': 0.010957665, '45': 0.006409571, '46': 0.004188655, '47': 10.0, '48': 5.0, '49': 2.0, '50': 2.0, '51': 1.0, 'receivingReceptions': 6.353417973, '54': 1.0, '56': 0.119723055, '57': 0.0037, '58': 7.688226444, '60': 8.53821278, '61': 54.24683453, '62': 0.032786623, '63': 0.000702, '66': 0.116859396, '67': 0.042757039, '68': 0.159616435, '70': 0.0538, '71': 0.023943942, 'lostFumbles': 0.077699264, '73': 0.077699264, '210': 1.0}}}
```

I noticed there's no season aggregate stat.  In the ESPN API I see what appears to be seasonal stats in the same stats array that the weekly is pulled from, and it has a `statSplitTypeId: 0` and `statSourceId: 0`.

A second issue is that the 2nd week isn't appearing in the stats response for any players yet.  I'm only seeing the first week, so I can't sum all of the weekly stats together.

**Comment by cwendt94** [comment, marked as answer]
Interesting, I recently added the `stats` field to the [player class](https://github.com/cwendt94/espn-api/wiki/Player-Class) and have not fully dived into the API response. The first use case of the stats field was in [box player class](https://github.com/cwendt94/espn-api/wiki/Box-Player-Class) to get the weekly projected and actual points.

I think it would be very useful having the players aggregated and projected season stats. I will look into it more in the next couple days and hopefully have a PR up sometime this weekend/next week!

**Reply by cwendt94** [reply]
@jcordell I did some digging into the stats and found a couple things

1. The season projected and aggregate stats are under key `0` however ESPN mixes 2019 and 2020 stats in that call. So right now it's showing 2019 aggregate and projected. I am going to update the logic to check that the stats are for the current year.
2. The basic league API call that gets every team's roster and player stats only returns the upcoming week, previous week and current season aggregate points so it won't have every weeks stats. If you want to get a players  stats for every week you can do something like this
```python
id = league.teams[0].roster[0].playerId
player = league.player_info(playerId=id)
``` 
This will get stats from every week for the player and more detailed data.

**Reply by cwendt94** [reply]
From issue #134 I am also going to add two new fields to the Player class `total_points` and `projected_total_points`.

**Reply by jcordell** [reply]
Awesome, thank you so much!  Looking forward to testing it out later this week

**Reply by cwendt94** [reply]
This update is now in package version v0.8.0

**Reply by jrdykstra** [reply]
It seems like in my league for years 2017 and under, the projected_total_points is 0.  Any thoughts on why?

```
league = League(league_id=league_id, year=2016, espn_s2=espn_s2, swid=swid)

id = league.draft[0].playerId
player = league.player_info(playerId=id)

print(player.name, player.total_points)
print(player.name, player.projected_total_points)

```
Output:
Antonio Brown 254.3
Antonio Brown 0

**Reply by cwendt94** [reply]
I checked the data ESPN is returning for league's older than 2018 and they are only returning the total season stats. Years greater than 2017 ESPN will return every week for a players stat.

I'm guessing they decided either to not save older years data or not include it in their API

**Reply by cwendt94** [reply]
`player_info` function is still there. What is the error you are seeing and version of the package?

**Reply by ghost** [reply]
> @jcordell I did some digging into the stats and found a couple things
> 
> 1. The season projected and aggregate stats are under key `0` however ESPN mixes 2019 and 2020 stats in that call. So right now it's showing 2019 aggregate and projected. I am going to update the logic to check that the stats are for the current year.
> 2. The basic league API call that gets every team's roster and player stats only returns the upcoming week, previous week and current season aggregate points so it won't have every weeks stats. If you want to get a players  stats for every week you can do something like this
> 
> ```python
> id = league.teams[0].roster[0].playerId
> player = league.player_info(playerId=id)
> ```
> 
> This will get stats from every week for the player and more detailed data.

So I assume we replace the 0 with a loop to get the data from all players, how do we get the weekly aggregation?

Thanks again for all your help!

**Reply by cwendt94** [reply]
So if you want to get the weekly points scored by your team/players you can use the `box_scores` [interface](https://github.com/cwendt94/espn-api/wiki/League-Class). Also each [player object](https://github.com/cwendt94/espn-api/wiki/Player-Class) has a `total_points` field for the whole year. The `total_points` field is filled out for every player on a team when the League is initialized.

The `player_info` call is mainly used if you want the break down of a players points per week or if they are not on anyone's team anymore.

**Reply by ghost** [reply]
I'm just trying to get all players stats aggregated by week from 2017 till now whether they were rostered or not. Tried using the Player class but getting this error:


  File "X", line 14, in <module>
    print(Player.playerId, Player.name, Player.stats)

AttributeError: type object 'Player' has no attribute 'playerId'

**Reply by cwendt94** [reply]
I'm not sure how you are accessing the player object but it should be through the League object.

To get every players stat starting from 2017 even if they were not rostered is a little tricky. First ESPN might rate limit your request because you will need to call their API for every player. Also ESPN does remove some of their player data after a certain year, I can't remember if it was 2017 or 2016. Here is some sample code of getting every players information for a given year
```python3
from espn_api.football import League
league = League(league_id=1245, year=2017)
for playerId in league.player_map:
   if  isinstance(playerId, int):
      player = league.player_info(playerId=playerId)
      if player:
         # now you have the player object just get the stats
         print(player.name, player.stats)   
```

**Comment by ghost** [comment]


**Reply by cwendt94** [reply]
Hmm that is very strange, `player_info` should be present for the Football League object. To check version, you can run this command
`pip show espn_api`

**Reply by ghost** [reply]
Ah, I didn't put the ff in front on espn_api. For pip sow ff_espn_api it shows 1.1.9.

**Reply by cwendt94** [reply]
Oh okay, so that is the old package. The new package is now call `espn_api`. So you will want to do a `pip install espn_api` for the latest!

**Comment by rab175** [comment]
I'm running into a curious issue where I'm getting no player info for Josh Allen. 

Either of these will return None: 

player = league.player_info(name='Josh Allen')
player = league.player_info(playerId= 3915239)

It also happens for AJ Green, but every other player seems to work fine. 

Anyone else experience this?

**Comment by ghost** [comment]
I think because Josh Allen is ambiguous. Also a defensive player.

I’m still trying to find the best way to get a stored data frame of each
players weekly stats over the last few years.

**Reply by rab175** [reply]
Thanks, seems like that's it! 

So, what I've determined is that the league.playermap is a dict that has both the player name and playerId, and has them both as a key and a value {name:playerId, playerId:name, etc.} . Since the the key can't repeat, the non-unique name (Josh Allen) only exists as a key once and its value is the defensive player's palyerId. However there are two 'Josh Allen' values with two different playerId keys.
